src/main/scala/lantern/ad_lms_vector.scala

```diff
@@ -203,22 +203,63 @@ trait TensorExp extends Dsl with Diff {
     * Tensor ops are defined in terms of primitive operations.
     */
   trait BackendNative extends Backend {
-    override def dot(x: Tensor, y: Tensor): Tensor = {
-      // TODO: (Fei Wang): only support 2D dot 1D and 1D dot 1D
-      val off = var_new(0)
-      val up = if (x.rank > 1) x.shape(0) else 1
-      val res = NewArray[Float](up)
-      for (j <- DataLoop(up)) {
+    // Compute vector-vector dot product, i.e. inner product.
+    // [V1] dot [V2] => [1] (scalar)
+    private def vvdot(x: Tensor, y: Tensor): Tensor = {
+      assert(x.shape(0) == y.shape(0))
+      val value = var_new(0.0f)
+      for (i <- DataLoop(x.shape.last)) {
+        value += x.data(i) * y.data(i)
+      }
+      val res = NewArray[Float](1)
+      res(0) = readVar(value)
+      Tensor(res, 1)
+    }
+
+    // Compute matrix-vector dot product.
+    // [M1 x M2] dot [M2] => [M1]
+    private def mvdot(x: Tensor, y: Tensor): Tensor = {
+      assert(x.shape(1) == y.shape(0))
+      val dim1 = x.shape(0)
+      val dim2 = x.shape(1)
+      val res = NewArray[Float](dim1)
+      for (i <- DataLoop(dim1)) {
         val value = var_new(0.0f)
-        for (i <- DataLoop(x.shape.last)) {
-          value += x.data(off) * y.data(i)
-          off += 1
+        for (j <- DataLoop(dim2)) {
+          value += x.data(i * dim2 + j) * y.data(j)
```

Comment:
I noticed other parts of the code use auxiliary index variables rather than doing index calculations inline:
```scala
var xIndex = var_new(0)
for (i <- DataLoop(dim1)) {
  for (j <- DataLoop(dim2)) {
    value += x.data(xIndex + j) * y.data(j)
    xIndex += dim2
    // As opposed to inline calculation:
    // `value += x.data(i * dim2 + j) * y.data(j)`
  }
}
```

I used inline calculations because I didn't want to optimize prematurely.

Reply:
It may be worth to see the assembly generated by GCC on the subject. It may matter for big matrix.